core/server/data/migrations/versions/4.35/2022-02-02-13-10-transform-specific-tiers-default-content-visibility.js

```diff
@@ -0,0 +1,107 @@
+const logging = require('@tryghost/logging');
+
+const {createTransactionalMigration} = require('../../utils');
+
+module.exports = createTransactionalMigration(
+    async function up(knex) {
+        logging.info('Checking default_content_visibility for specific tiers');
+
+        const settings = await knex('settings')
+            .select()
+            .whereIn('key', ['default_content_visibility', 'default_content_visibility_tiers']);
+        const contentVisibilitySetting = settings.find(d => d.key === 'default_content_visibility');
+        const visibilityTiersSetting = settings.find(d => d.key === 'default_content_visibility_tiers');
+        if (!contentVisibilitySetting || !contentVisibilitySetting.value) {
+            logging.warn('No default_content_visibility setting found.');
+            return;
+        }
+
+        if (!visibilityTiersSetting) {
+            logging.warn('No default_content_visibility_tiers setting found.');
+            return;
+        }
+
+        const contentVisibility = contentVisibilitySetting.value;
+
+        // Transform visibility to tiers when stored as nql string
+        if (!['public', 'members', 'paid'].includes(contentVisibility)) {
+            // fetch product slugs from nql filter
+            const productSlugs = contentVisibility.split(',').map((segment) => {
```

Comment:
Should we be validating the format of this string? ATM it could be any arbitrary NQL filter right?


From what I can tell we're expecting something like `product:slug,product:other-slug,product:etc`

I think a regex like `/^(?:product:[\w\-]+,?)+$/` should work to test this, and we can just use default values if the filter doesn't match - what do you think?

Reply:
@allouis 🤔 Yea that's a good point, while ideally sites should not have any arbitrary NQL filter other than products or label, it's a good idea to handle the invalid ones and fallback to `paid` as default instead which is the highest level instead.

Reply:
Updated ✔️